### PR TITLE
Rest api update

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will update the rest API, and add a websocket

### Proposed changes

- update rest API
- add websocket server
- fix a few issues with the webserver, and make it stoppable
- add a test suite for the webserver 
- update docs

### CODE REVIEW

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/webserver/webServerWebSocketTests.cpp  24
- src/helics/apps/helicsWebServer.cpp  2
- tests/helics/webserver/helics_webserver_tests.cpp  1
- tests/helics/webserver/webServerHttpTests.cpp  18
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5315113&bid=17394893)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/webserver/helics_webserver_tests.cpp  1
- tests/helics/webserver/webServerHttpTests.cpp  18
- tests/helics/webserver/webServerWebSocketTests.cpp  24
- src/helics/apps/helicsWebServer.cpp  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5315113&bid=17394893)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/webserver/helics_webserver_tests.cpp  1
- src/helics/apps/helicsWebServer.cpp  2
- tests/helics/webserver/webServerHttpTests.cpp  16
- tests/helics/webserver/webServerWebSocketTests.cpp  23
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5315113&bid=17394893)

File: docs/user-guide/simultaneous_cosimulations.md [hunk 1]
```diff
@@ -1,67 +1,2 @@
-# Simultaneous Co-simulations
```
Comment (nightlark) [inline docs/user-guide/simultaneous_cosimulations.md:1]: Was removing the rest of this guide the intended?

Comment (phlptp) [reply docs/user-guide/simultaneous_cosimulations.md:1]: I changed the name of the page to be better but I wanted the link to remain so this is a just a way to redirect.  All the content is now in webserver.md

Comment (nightlark) [reply docs/user-guide/simultaneous_cosimulations.md:1]: What about the sections on setting port numbers and use of keys?

Comment (nightlark) [reply docs/user-guide/simultaneous_cosimulations.md:1]: I think those are still relevant for multi-user systems like HPC nodes.

Comment (phlptp) [reply docs/user-guide/simultaneous_cosimulations.md:1]: You are correct I think I removed the wrong file,  I will correct that in the next commit

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/webserver/webServerWebSocketTests.cpp  23
- tests/helics/webserver/webServerHttpTests.cpp  16
- tests/helics/webserver/helics_webserver_tests.cpp  1
- src/helics/apps/helicsWebServer.cpp  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5315113&bid=17394893)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/apps/helicsWebServer.cpp  2
- tests/helics/webserver/helics_webserver_tests.cpp  1
- tests/helics/webserver/webServerWebSocketTests.cpp  23
- tests/helics/webserver/webServerHttpTests.cpp  16
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5315113&bid=17394893)

File: docs/user-guide/running_interaction.md [hunk 2]
```diff
@@ -1,138 +1,2 @@
...
-
-## Queries
-
-Currently any query is accessible through this interface.  Queries have a target and a query.  The target is some named object in the federation and the query is a question.  The available queries are listed [HERE](queries.md).  More are expected to be added.  
-
-## Notes
-This is an experimental interface and is subject to change and improvements.  Suggestions are welcome.  Also a websocket version of this is expected in the next release to allow websockets to connect.  
-
-The most likely use case for this will be as a component for a more sophisticated control interface, so a more user friendly setup will be using the webserver as a back-end for control, debugging, information, and visualization of a running co-simulation.  
+# webserver
```
Comment (nightlark) [inline docs/user-guide/running_interaction.md:1]:
```suggestion
# Webserver
```

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/webserver/webServerHttpTests.cpp  16
- tests/helics/webserver/webServerWebSocketTests.cpp  23
- tests/helics/webserver/helics_webserver_tests.cpp  1
- src/helics/apps/helicsWebServer.cpp  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5315113&bid=17394893)

File: docs/user-guide/webserver.md [hunk 3]
```diff
@@ -0,0 +1,228 @@
+# Interacting with a Running Simulation
+
+Starting in HELICS 2.4 there is a webserver that can be run with the broker_server  or helics_broker_server.  This requires using a boost version >=1.70.  The Webserver can be disabled by the `HELICS_DISABLE_BOOST=ON` or `HELICS_DISABLE_WEBSERVER=ON` options being set.
```
Comment (kdheepak) [inline docs/user-guide/webserver.md:3]: broker_server or helics_broker_server

File: docs/user-guide/webserver.md [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+making an invalid query will produce
+`http://localhost/brokerA/i_dont_care`  -> `#invalid`
+
+Queries can be make in a number of different formats, the following are equivalent
+-   `http://localhost/brokerA/publications`
+-   `http://localhost/brokerA`
+-   `http://localhost/brokerA?query=publications`
+-   `http://localhost/brokerA/publications`
+-   `http://localhost/brokerA/root/publications`
+-   `http://localhost?broker=brokerA&query=publications&target=root`
```
Comment (kdheepak) [inline docs/user-guide/webserver.md:203]: Should this be:

```
http://localhost/?broker=brokerA&query=publications&target=root
```

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/webserver/webServerHttpTests.cpp  16
- tests/helics/webserver/webServerWebSocketTests.cpp  23
- src/helics/apps/helicsWebServer.cpp  2
- tests/helics/webserver/helics_webserver_tests.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5315113&bid=17394893)

**Comment by kdheepak** [comment]
I'm excited to try out this feature!

File: docs/user-guide/webserver.md (hunk 4 above)
Comment (phlptp) [reply docs/user-guide/webserver.md:203]: I am not sure if that matters or not

**Pull request merged**